core/src/main/scala/quasar/destination/gbq/GBQFlow.scala

```diff
@@ -269,18 +267,40 @@ final class GBQFlow[F[_]: Concurrent](
     })
   }
 
-  private def upload(bytes: Stream[F, Byte], uploadLocation: Uri): Stream[F, Unit] = {
-    val destReq = Request[F](Method.PUT, uploadLocation)
-        .putHeaders(Header("Host", "www.googleapis.com"))
-        .withContentType(`Content-Type`(MediaType.application.`x-www-form-urlencoded`))
-        .withEntity(bytes)
-    client.stream(destReq) evalMap { resp =>
-      resp.status match {
-        case Status.Ok => ().pure[F]
-        case _ => ApplicativeError[F, Throwable].raiseError[Unit](
-            new RuntimeException(s"Upload failed: ${resp.status.reason}" ))
-      }
-    }
+  private def upload(bytes: Stream[F, Byte], uploadLocation: F[Either[InitializationError[Json], Uri]]): Stream[F, Unit] = {
+
+    bytes.pull.uncons.flatMap {
```

Comment:
⛳️ You could use `pull.peek` here instead so you don't have to concat the stream again.